helix-term/src/ui/context.rs

```diff
@@ -0,0 +1,416 @@
+        query,
+        start_node.unwrap_or_else(|| tree.root_node()),
+        RopeProvider(text),
+    );
+
+    for matched_node in query_nodes {
+        // find @context.params nodes
+        let node_byte_range = get_context_paired_range(
+            &matched_node,
+            start_index,
+            end_index,
+            top_first_byte,
+            last_scan_byte,
+        );
+
+        for node in matched_node.nodes_for_capture_index(start_index) {
+            if (!node.byte_range().contains(&last_scan_byte)
+                || !node.byte_range().contains(&top_first_byte))
+                && node.start_position().row != anchor_line + result.len()
+                && node_byte_range.is_none()
+            {
+                continue;
+            }
+
+            result.push(StickyNode {
+                line: node.start_position().row,
+                visual_line: 0,
+                byte_range: node_byte_range
+                    .as_ref()
+                    .unwrap_or(&(node.start_byte()..node.end_byte()))
+                    .clone(),
+                indicator: None,
+                anchor: view.offset.anchor,
+                has_context_end: node_byte_range.is_some(),
+            });
+        }
+    }
+    // result should be filled by now
+    if result.is_empty() {
+        if !cached_nodes.is_empty() {
+            return Some(cached_nodes);
+        }
+
+        return None;
+    }
+
+    let mut res = {
+        cached_nodes.append(&mut result);
+        cached_nodes
+    };
+
+    // Order of commands is important here
+    res.sort_unstable_by(|lhs, rhs| lhs.line.cmp(&rhs.line));
+    res.dedup_by(|lhs, rhs| lhs.line == rhs.line);
+
+    // always cap the maximum amount of sticky contextes to 1/3 of the viewport
+    // unless configured otherwise
+    let max_lines = config.sticky_context.max_lines as u16;
+    let max_nodes_amount = max_lines.min(viewport.height / 3) as usize;
+
+    let skip = res.len().saturating_sub(max_nodes_amount);
+
+    res = res
+        .iter()
+        // only take the nodes until 1 / 3 of the viewport is reached or the maximum amount of sticky nodes
+        .skip(skip)
+        .enumerate()
+        .take_while(|(i, _)| {
+            *i + Into::<usize>::into(config.sticky_context.indicator) != visual_cursor_row as usize
+        }) // also only nodes that don't overlap with the visual cursor position
+        .map(|(i, node)| {
+            let mut new_node = node.clone();
+            new_node.visual_line = i as u16;
+            new_node
+        })
+        .collect();
+
+    if config.sticky_context.indicator {
+        let str = "─".repeat(viewport.width as usize);
+
+        res.push(StickyNode {
+            line: usize::MAX,
+            visual_line: res.len() as u16,
+            byte_range: 0..0,
+            indicator: Some(str),
+            anchor: view.offset.anchor,
+            has_context_end: false,
+        });
+    }
+
+    Some(res)
+}
```

Comment:
**Suggestion:** The function `calculate_sticky_nodes` is quite long and complex. Consider breaking it down into smaller, more manageable functions. This will improve readability and maintainability.
```suggestion
// This is a high-level example. The actual implementation may vary.
pub fn calculate_sticky_nodes(
    nodes: &Option<Vec<StickyNode>>,
    doc: &Document,
    view: &View,
    config: &helix_view::editor::Config,
    cursor_cache: &Option<Option<Position>>,
) -> Option<Vec<StickyNode>> {
    let cursor_cache = validate_cursor_cache(cursor_cache)?;
    let (syntax, tree, text, viewport, cursor_byte) = initialize_variables(doc, view)?;
    let (anchor_line, visual_cursor_row, top_first_byte, last_scan_byte) = calculate_initial_values(view, nodes, text, cursor_byte)?;
    let cached_nodes = get_cached_nodes(nodes, view, visual_cursor_row)?;
    let (start_byte_range, start_byte) = calculate_start_byte_range(tree, cached_nodes, last_scan_byte)?;
    let start_node = get_start_node(tree, start_byte_range)?;
    let context_nodes = get_context_nodes(doc)?;
    let (start_index, end_index) = get_indices(context_nodes)?;
    let mut result = calculate_result(start_node, tree, text, top_first_byte, last_scan_byte, context_nodes, start_index, end_index, anchor_line, view)?;
    result = finalize_result(result, viewport, config)?;
    Some(result)
}
```